### PR TITLE
[JOB LAUNCHER SERVER] API KEY

### PR DESCRIPTION
**Pull request by sarthakvijayvergiya**
## Description

This pull request introduces the functionality to validate an API key for a user. The changes ensure that the API key is securely hashed and can be verified to determine its authenticity. This is crucial for the secure operation of our API, ensuring that only authorized users can access certain endpoints.

## Summary of changes

Updated .env.example in the Job Launcher Server to include essential environment variables for API key validation.
Enhanced env.ts in the common config to handle secure API key processing configurations.
Added new error definitions in errors.ts for API key validation failures.
Created apikey.ts in the common decorators to facilitate API key handling in the application.
Updated index.ts in both decorators and guards directories to include new API key-related functionalities.
Created apikey.auth.spec.ts and apikey.auth.ts in common guards for testing and implementing API key authentication.
Enhanced jwt.auth.ts to integrate JWT authentication with API key validation.
Added cryptographic utilities in crypto.ts for secure API key handling.
Configured database.module.ts to support API key management.
Created migration 1699446310308-addApiKeyEntity.ts in the database for API key entity setup.
Introduced apikey.entity.ts and apikey.repository.ts in the auth module for managing API keys at the database level.
Integrated API key management endpoints and logic in auth.controller.ts.
Updated auth.module.ts to include new API key management services and controllers.
Upgraded auth.service.ts to handle API key hashing, verification, and security processes.
Ensured compatibility with API key authentication in job.controller.ts and job.module.ts.
Modified user.entity.ts to include API key-related fields and associations.

## Related issues
[Keywords for linking issues](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/using-keywords-in-issues-and-pull-requests)

Fixes humanprotocol/human-protoco#948
https://github.com/humanprotocol/human-protocol/issues/948

## Operational checklist

- [ ] All new functionality is covered by tests
- [ ] Any related documentation has been changed or added

### CODE REVIEW

File: packages/apps/job-launcher/server/src/modules/user/user.service.ts [hunk 1]
```diff
@@ -100,4 +101,24 @@ export class UserService {
       currency: Currency.USD,
     };
   }
+
+  async createOrUpdateAPIKey(userId: number): Promise<string> {
+    const salt = crypto.randomBytes(16).toString('hex');
+    const apiKey = crypto.randomBytes(32).toString('hex');
+    const hashedAPIKey = crypto.pbkdf2Sync(apiKey, salt, 1000, 64, `sha512`).toString(`hex`);
```
Comment (mrhouzlane) [inline packages/apps/job-launcher/server/src/modules/user/user.service.ts:108]: 1000 and 64 can be added using configService.

**Review by mrhouzlane — CHANGES_REQUESTED** [review]
Also API Key should be valid indefinitely or it expires here ?

File: packages/apps/job-launcher/server/src/modules/user/user.entity.ts [hunk 2]
```diff
@@ -39,4 +39,33 @@ export class UserEntity extends BaseEntity implements IUser {
 
   @OneToMany(() => PaymentEntity, (payment) => payment.user)
   public payments: PaymentEntity[];
+
+  @OneToOne(() => ApiKeyEntity, apiKey => apiKey.user)
+  public apiKey: ApiKeyEntity;
+
 }
+
+@Entity({ schema: NS, name: 'api_keys' })
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/modules/user/user.entity.ts:48]: ApiKeyEntity should be in auth module

**Review by portuu3 — CHANGES_REQUESTED** [review]
@sarthakvijayvergiya you should include tests

**Review by portuu3 — CHANGES_REQUESTED** [review]
migrations for API key entity are missing

File: packages/apps/job-launcher/server/src/modules/job/job.controller.ts [hunk 3]
```diff
@@ -37,6 +37,7 @@ import { Role } from '../../common/enums/role';
 export class JobController {
   constructor(private readonly jobService: JobService) {}
 
+  @UseGuards(ApiKeyGuard)
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/modules/job/job.controller.ts:40]: Can this endpoint be called with both API key and jwt?

File: packages/apps/job-launcher/server/src/common/guards/apikey.auth.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import { Injectable, ExecutionContext, UnauthorizedException, CanActivate } from '@nestjs/common';
+import { Reflector } from '@nestjs/core';
+import { AuthService } from 'src/modules/auth/auth.service';
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/common/guards/apikey.auth.ts:3]: this route needs to be relative

File: packages/apps/job-launcher/server/src/database/database.module.ts [hunk 5]
```diff
@@ -12,6 +12,7 @@ import { TypeOrmLoggerModule, TypeOrmLoggerService } from './typeorm';
 import { JobEntity } from '../modules/job/job.entity';
 import { PaymentEntity } from '../modules/payment/payment.entity';
 import { ConfigNames } from '../common/config';
+import { ApiKeyEntity } from 'src/modules/auth/apikey.entity';
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/database/database.module.ts:15]: this route needs to be relative

File: packages/apps/job-launcher/server/src/modules/auth/auth.controller.ts [hunk 6]
```diff
@@ -23,10 +28,13 @@ import {
 import { AuthService } from './auth.service';
 import { JwtAuthGuard } from '../../common/guards';
 import { RequestWithUser } from '../../common/types';
+import { ErrorAuth } from 'src/common/constants/errors';
```
Comment (portuu3) [inline packages/apps/job-launcher/server/src/modules/auth/auth.controller.ts:31]: this route needs to be relative

**Review by portuu3 — CHANGES_REQUESTED** [review]

File: packages/apps/job-launcher/server/src/modules/user/user.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (sarthakvijayvergiya) [reply packages/apps/job-launcher/server/src/modules/user/user.service.ts]: I like that, included it.

Comment (sarthakvijayvergiya) [reply packages/apps/job-launcher/server/src/modules/user/user.service.ts]: yes, added in config service!!

File: packages/apps/job-launcher/server/src/modules/user/user.entity.ts (hunk 2 above)
Comment (sarthakvijayvergiya) [reply packages/apps/job-launcher/server/src/modules/user/user.entity.ts]: yep, moved it!!

File: packages/apps/job-launcher/server/src/common/guards/apikey.auth.ts (hunk 4 above)
Comment (sarthakvijayvergiya) [reply packages/apps/job-launcher/server/src/common/guards/apikey.auth.ts]: fixed!

File: packages/apps/job-launcher/server/src/modules/user/user.controller.ts [hunk 7]
```diff
@@ -38,4 +41,23 @@ export class UserController {
...
+  @HttpCode(HttpStatus.CREATED)
+  public async createOrUpdateAPIKey(
+    @Request() req: RequestWithUser,
+  ): Promise<{ apiKey: string }> {
+    try {
+      const apiKey = await this.userService.createOrUpdateAPIKey(req.user.id);
+      return { apiKey };
+    } catch (e) {
+      this.logger.log(
+        e.message,
```
[the remark this answers is not in the document]
Comment (sarthakvijayvergiya) [reply packages/apps/job-launcher/server/src/modules/user/user.controller.ts]: no, it's a error message!

File: packages/apps/job-launcher/server/src/database/database.module.ts (hunk 5 above)
Comment (sarthakvijayvergiya) [reply packages/apps/job-launcher/server/src/database/database.module.ts]: fixed!

**Pull request merged**